peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+					// command.
+					p.PushRejectMsg("malformed",
+						wire.RejectMalformed, errMsg,
+						nil, true)
+				}
+
+			}
+			break out
+		}
+		p.StatsMtx.Lock()
+		p.lastRecv = time.Now()
+		p.StatsMtx.Unlock()
+
+		// Ensure version message comes first.
+		if vmsg, ok := rmsg.(*wire.MsgVersion); !ok && !p.VersionKnown() {
+			errStr := "A version message must precede all others"
+			p.LogError(errStr)
+
+			// Push a reject message and wait for the message to be
+			// sent before disconnecting.
+			p.PushRejectMsg(vmsg.Command(), wire.RejectMalformed,
+				errStr, nil, true)
+			break out
+		}
+
+		// Handle each supported message type.
+		switch msg := rmsg.(type) {
+		case *wire.MsgVersion:
+			p.handleVersionMsg(msg)
+			for key, listener := range p.versionMsgListeners {
+				log.Tracef("Running %s listener %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgVerAck:
+			p.StatsMtx.Lock()
+			versionSent := p.versionSent
+			verAckReceived := p.verAckReceived
+			p.StatsMtx.Unlock()
+
+			if !versionSent {
+				log.Infof("Received 'verack' from peer %v "+
+					"before version was sent -- disconnecting", p)
+				break out
+			}
+			if verAckReceived {
+				log.Infof("Already received 'verack' from "+
+					"peer %v -- disconnecting", p)
+				break out
+			}
+			p.verAckReceived = true
+
+		case *wire.MsgGetAddr:
+			for key, listener := range p.getAddrMsgListeners {
+				log.Tracef("Running %s listener %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgAddr:
+			for key, listener := range p.addrMsgListeners {
+				log.Tracef("Running %s listener %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgPing:
+			p.handlePingMsg(msg)
+
+		case *wire.MsgPong:
+			p.handlePongMsg(msg)
+
+		case *wire.MsgAlert:
+			// Intentionally ignore alert messages.
+			//
+			// The reference client currently bans peers that send
+			// alerts not signed with its key.  We could verify
+			// against their key, but since the reference client
+			// is currently unwilling to support other
+			// implementions' alert messages, we will not relay
+			// theirs.
+
+		case *wire.MsgMemPool:
+			for key, listener := range p.memPoolMsgListeners {
+				log.Tracef("Running %s listener %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgTx:
+			for key, listener := range p.txMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgBlock:
+			for key, listener := range p.blockMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg, buf)
+			}
+
+		case *wire.MsgInv:
+			for key, listener := range p.invMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgHeaders:
+			for key, listener := range p.headersMsgListeners {
+				log.Tracef("Running %s listener for %s", key, p)
+				listener(p, msg)
+			}
+
+		case *wire.MsgNotFound:
+			// TODO(davec): Ignore this for now, but ultimately
```

Comment:
Even though it's not used yet, we should expose a listener for this message.  It will be needed when we implement concurrent block downloading.  Also, other callers might need to react to it.

The comment can then be moved into the registration code in btcd proper.